onnx/numpy_helper.py

```diff
@@ -397,6 +470,62 @@ def from_array(arr: np.ndarray, name: str | None = None) -> TensorProto:
     return tensor
 
 
+def from_array(tensor: np.ndarray, name: str | None = None) -> TensorProto:
+    """Converts an array into a TensorProto including
+    supported type defined in :mod:`onnx.custom_element_types`.
+
+    Args:
+        tensor: a numpy array.
+        name: (optional) the name of the tensor.
+
+    Returns:
+        TensorProto: the converted tensor def.
+    """
+    if not isinstance(tensor, np.ndarray):
+        return _from_array(tensor, name)
+
+    dt = tensor.dtype
+    if dt == custom_np_types.float8e4m3fn and dt.descr[0][0] == "e4m3fn":
+        to = TensorProto.FLOAT8E4M3FN
+        dt_to = np.uint8  # type: ignore[assignment]
+    elif dt == custom_np_types.float8e4m3fnuz and dt.descr[0][0] == "e4m3fnuz":
+        to = TensorProto.FLOAT8E4M3FNUZ
+        dt_to = np.uint8  # type: ignore[assignment]
+    elif dt == custom_np_types.float8e5m2 and dt.descr[0][0] == "e5m2":
+        to = TensorProto.FLOAT8E5M2
+        dt_to = np.uint8  # type: ignore[assignment]
+    elif dt == custom_np_types.float8e5m2fnuz and dt.descr[0][0] == "e5m2fnuz":
+        to = TensorProto.FLOAT8E5M2FNUZ
+        dt_to = np.uint8  # type: ignore[assignment]
+    elif dt == custom_np_types.bfloat16 and dt.descr[0][0] == "bfloat16":
+        to = TensorProto.BFLOAT16
+        dt_to = np.uint16  # type: ignore[assignment]
+    elif dt == custom_np_types.int4 and dt.descr[0][0] == "int4":
+        to = TensorProto.INT4
+        dt_to = np.int8  # type: ignore[assignment]
+    elif dt == custom_np_types.uint4 and dt.descr[0][0] == "uint4":
+        to = TensorProto.UINT4
+        dt_to = np.uint8  # type: ignore[assignment]
+    else:
+        return _from_array(tensor, name)
+
+    if to in (TensorProto.UINT4, TensorProto.INT4):
+        value = tensor.astype(dt_to).ravel()
+        buffer = (value[::2] & 0x0F) + (value[1::2] << 4)
```

Comment:
Do we need to handle odd-sized arrays? https://github.com/microsoft/onnxscript/blob/1aa7a7017225f95621e9a2008041056661ff7977/onnxscript/ir/_type_casting.py#L23-L25

Reply:
We probably should.

Reply:
I raised an exception. Odd size should not happen.